recipe/bld.bat

```diff
@@ -1,20 +1,36 @@
-mkdir "%SRC_DIR%"\build
-pushd "%SRC_DIR%"\build
+@echo on
 
-if NOT "%cuda_compiler_version%"=="None" (
+mkdir %SRC_DIR%\build
+pushd %SRC_DIR%\build
+
+:: CCCL has different logic depending on the `__cplusplus` macro.
+:: However Visual Studio leaves `__cplusplus` as `199711L` by default.
+:: As a result code paths needing new C++ standard won't be run.
+:: This happens in spite of the fact that VS 2019 & 2022 default to C++14.
+:: To make sure `__cplusplus` matches, we pass an additional flag to VC.
+:: Ideally other tooling would do this for us, for now we do it ourselves.
+::
+:: https://learn.microsoft.com/en-us/cpp/build/reference/zc-cplusplus?view=msvc-160
+:: https://learn.microsoft.com/en-us/cpp/build/reference/zc-cplusplus?view=msvc-170
+:: https://github.com/NVIDIA/cccl/blob/82a3ed0282893d6316abde04394f59a9a37a3747/libcudacxx/include/nv/target#L27-L30
+:: https://github.com/NVIDIA/cutlass/issues/1403#issuecomment-2130383392
+:: https://gitlab.kitware.com/cmake/cmake/-/issues/18837
+set "CXXFLAGS=%CXXFLAGS% /Zc:__cplusplus"
```

Comment:
Ok did a bit of a deep dive and have a theory of what is going on. Tried to explain this in the comment above

Basically some C++ detection logic in libraries ([like CCCL]( https://github.com/NVIDIA/cccl/blob/82a3ed0282893d6316abde04394f59a9a37a3747/libcudacxx/include/nv/target#L27-L30 )) doesn't work correct with Visual Studio due to its default behavior. Have added a flag to Visual Studio to make it conform more to typical C++ compiler behavior (setting `__cplusplus` correctly)

Based on similar experiences from Visual Studio users with CUTLASS ( https://github.com/NVIDIA/cutlass/issues/1403#issuecomment-2130383392 ), who saw a similar error to us and [fixed it with this flag]( https://github.com/NVIDIA/cutlass/blob/56b46e2d13875b46b8f6a03f9f5ac91e2bfdc01a/CMakeLists.txt#L442-L452 ), am hopeful this will help in our case too

Ideally other tooling would handle this for us, but it appears [CMake hasn't done this yet]( https://gitlab.kitware.com/cmake/cmake/-/issues/18837 ) and Visual Studio continues to require users to add this flag if they want a correct `__cplusplus` value. So for now adding to this build script. If this works, we may want to add this to XGBoost as well

Reply:
Have raised an upstream CCCL issue: https://github.com/NVIDIA/cccl/issues/2072

Though after looking more closely at the code there, they are probably ok with their current approach. This would be a potential simplification for them